### PR TITLE
RHCLOUD-29011 Harden the delivery guarantee

### PR DESCRIPTION
**Pull request by gwenneg**

### CODE REVIEW

File: src/main/java/com/redhat/cloud/notifications/GwResource.java [hunk 1]
```diff
@@ -125,17 +134,32 @@ public Response forward(@NotNull @Valid RestAction ra) {
...
-        return Response.ok().build();
+        CompletableFuture<Void> callback = new CompletableFuture<>();
+        emitter.send(buildMessageWithId(serializedAction, callback));
+        try {
+            callback.get(callbackTimeout, TimeUnit.SECONDS);
+            forwardedActions.increment();
+            return Response.ok().build();
+        } catch (Exception e) {
+            Log.error("Message delivery to Kafka failed", e);
+            return Response.status(SERVICE_UNAVAILABLE).build();
```
Comment (gwenneg) [inline src/main/java/com/redhat/cloud/notifications/GwResource.java:146]: @jharting Does returning 503 look good to you if we failed to deliver the Kafka message?

File: src/main/java/com/redhat/cloud/notifications/GwResource.java [hunk 2]
```diff
@@ -125,17 +134,32 @@ public Response forward(@NotNull @Valid RestAction ra) {
...
-        return Message.of(payload).addMetadata(metadata);
+        return Message.of(payload)
+                .addMetadata(metadata)
+                .withAck(() -> {
+                    callback.complete(null);
+                    return CompletableFuture.completedFuture(null);
+                }).withNack(reason -> {
+                    callback.completeExceptionally(reason);
+                    return CompletableFuture.completedFuture(null);
+                });
```
Comment (gwenneg) [inline src/main/java/com/redhat/cloud/notifications/GwResource.java:163]: As weird as it seems, this is [inspired from SmallRye Reactive Messaging itself](https://github.com/smallrye/smallrye-reactive-messaging/blob/e59e7cff9962c0cdf4a7f6de859add786a2a6a23/smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/providers/extension/EmitterImpl.java#L30-L38).

**Comment by gwenneg** [comment]
I'm trying to find a way to test the new code.

**Comment by gwenneg** [comment]
After trying a lot of approaches, I couldn't find a way to test this from unit tests... `InMemoryConnector` can't be used to mock failures and it seems impossible to inject a mocked `Emitter`.

**Review by MikelAlejoBR — APPROVED** [review]
Interesting way of using the callback. LGTM!

**Pull request merged**